Tweaks to question media escaping to fix media output

roundhill opened the pull request:
Fixes #2443

#### Changes proposed in this Pull Request:

#2443 was caused by using `wp_kses_post` on the output from `wp_video_shortcode`, which was stripping out the `<source>` tag that is necessary for a browser to load a video.

I've also removed another usage of `wp_kses_post` in `the_question_media()` which seemed redundant because the output was already being escaped in `get_the_question_media()`. This required a `phpcs` ignore though.

I started to write a test but wasn't sure how I could add an actual media attachment for testing the output 🤔 Suggestions welcome!

#### Testing instructions:

* Create some quiz questions and an image, audio and video file to each.
* Verify that the content appears properly in the question on the front-end of the site.

#### Proposed changelog entry for your changes:
Fixes video playback for question media attachments.

donnapep commented:
Would you mind confirming what is being entered into the question description field? I was trying to reproduce this issue prior to this fix, but when I dropped a link from the media library into the field, the video played without issue.

donnapep commented:
> I've also removed another usage of `wp_kses_post` in `the_question_media()` which seemed redundant because the output was already being escaped in `get_the_question_media()`. This required a `phpcs` ignore though.

WordPress encourages late escaping, meaning that escaping should ideally occur at the time that text is echoed to the screen. This avoids having to trace back through the code to verify that the text has indeed been escaped elsewhere. For this reason, unless there's an issue with escaping it twice, it may be better to leave the late escaping in.

roundhill commented:
> WordPress encourages late escaping, meaning that escaping should ideally occur at the time that text is echoed to the screen. This avoids having to trace back through the code to verify that the text has indeed been escaped elsewhere. For this reason, unless there's an issue with escaping it twice, it may be better to leave the late escaping in.

This was also removing the `<source>` tag in my testing which is why I removed it and added the comment for future devs. It appears to be stripped by `wp_kses_post()`. Maybe we can add an exception for that tag somehow and keep the escape function there?

> Would you mind confirming what is being entered into the question description field? I was trying to reproduce this issue prior to this fix, but when I dropped a link to an mp4 file from the media library into the field, the video played without issue.

I was manually attaching the file (not in the media library yet). It was this file: https://cloudup.com/ckok782lxQf

<img width="255" alt="Screen Shot 2020-02-13 at 8 39 09 AM" src="https://user-images.githubusercontent.com/789137/74451161-71a71100-4e3c-11ea-813e-062de41c8669.png">

renatho commented:
> This was also removing the <source> tag in my testing which is why I removed it and added the comment for future devs. It appears to be stripped by wp_kses_post(). Maybe we can add an exception for that tag somehow and keep the escape function there?

Maybe using the `wp_kses` and as allowed sent the `wp_kses_allowed_html( 'post' )` with the additional `<source>`?

roundhill commented:
> Maybe using the `wp_kses` and as allowed use the `wp_kses_allowed_html( 'post' )` concatenating the additional `<source>`?

I'll try that! Thanks.

roundhill commented:
OK I whitelisted the `source` tag with the `src` and `type` attributes and merged it with the `wp_kses_post` array. This way we can still escape at the moment of output 👍 

https://github.com/Automattic/sensei/pull/2886/commits/159a67f7056666af09447fcc6e299d273d2f6993

donnapep commented on includes/class-sensei-question.php:
```diff
@@ -641,17 +641,34 @@ public static function get_the_question_media( $question_id ) {
+	private static function question_media_kses_post( $source_string ) {
+		$source_tag = array(
+			'source' => array(
+				'type' => true,
```
I believe that both the `true` values here should be `array()`. 🤔

roundhill replied:
`array()` does also work, but I was following the format defined here: https://core.trac.wordpress.org/browser/tags/5.3/src/wp-includes/kses.php#L61

donnapep replied:
Good to know. I think I like `true` better. 🙂
